File: frontend/src/types/stock.ts
/**
 * 股票基本信息
 */
export interface Stock {
  symbol: string // 股票代码
  name: string // 股票名称
  market: 'A股' | '港股' | '美股' // 所属市场
  current_price: number // 当前价格
  change_amount: number // 涨跌额
  change_rate: number // 涨跌幅 (%)
  volume?: number // 成交量
  turnover?: number // 成交额
  high?: number // 最高价
  low?: number // 最低价
  open?: number // 开盘价
  prev_close?: number // 昨收价
  market_cap?: number // 总市值
}

/**
 * 股票详细信息
 */
export interface StockDetail extends Stock {
  pe_ratio?: number // 市盈率
  pb_ratio?: number // 市净率
  dividend_yield?: number // 股息率
  total_shares?: number // 总股本
  float_shares?: number // 流通股本
  industry?: string // 所属行业
  description?: string // 公司简介
}

/**
 * 股票查询参数
 */
export interface StockQueryParams {
  keyword?: string // 搜索关键词（代码/名称/拼音）
  market?: 'A股' | '港股' | '美股' // 市场筛选
  page?: number
  page_size?: number
}

/**
 * 股票查询响应
 */
export interface StockQueryResponse {
  total: number
  page: number
  page_size: number
  list: Stock[]
}

/**
 * K线数据点
 */
export interface KLineDataPoint {
  timestamp: string // 时间戳
  open: number // 开盘价
  high: number // 最高价
  low: number // 最低价
  close: number // 收盘价
  volume: number // 成交量
}

/**
 * K线数据
 */
export interface KLineData {
  symbol: string
  name: string
  period: 'day' | 'week' | 'month' // 周期
  data: KLineDataPoint[]
}

File: backend/app/schemas/stock.py
"""
Stock Schemas
"""

from pydantic import BaseModel, Field
from typing import Optional
from decimal import Decimal
from datetime import datetime


class StockBase(BaseModel):
    """Base stock schema"""

    symbol: str = Field(..., max_length=20, description="股票代码")
    name: str = Field(..., max_length=100, description="股票名称")
    market: str = Field(..., description="市场类型: A股/港股/美股")


class StockQuote(StockBase):
    """Stock quote schema"""

    current_price: Decimal = Field(..., description="当前价格")
    change_amount: Decimal = Field(..., description="涨跌额")
    change_rate: Decimal = Field(..., description="涨跌幅 (%)")
    open_price: Optional[Decimal] = Field(None, description="开盘价")
    high_price: Optional[Decimal] = Field(None, description="最高价")
    low_price: Optional[Decimal] = Field(None, description="最低价")
    prev_close: Optional[Decimal] = Field(None, description="昨收价")
    volume: Optional[Decimal] = Field(None, description="成交量")
    turnover: Optional[Decimal] = Field(None, description="成交额")
    amplitude: Optional[Decimal] = Field(None, description="振幅 (%)")
    turnover_rate: Optional[Decimal] = Field(None, description="换手率 (%)")
    pe_ratio: Optional[Decimal] = Field(None, description="市盈率")
    pb_ratio: Optional[Decimal] = Field(None, description="市净率")
    market_cap: Optional[Decimal] = Field(None, description="总市值")
    circulating_market_cap: Optional[Decimal] = Field(None, description="流通市值")
    quote_time: datetime = Field(..., description="行情时间")


class StockInfo(StockBase):
    """Stock basic information"""

    stock_id: int = Field(..., description="股票ID")
    industry: Optional[str] = Field(None, description="所属行业")
    sector: Optional[str] = Field(None, description="所属板块")
    list_date: Optional[datetime] = Field(None, description="上市日期")
    is_delisted: bool = Field(default=False, description="是否退市")

    class Config:
        from_attributes = True


class StockSearch(BaseModel):
    """Stock search result"""

    symbol: str = Field(..., description="股票代码")
    name: str = Field(..., description="股票名称")
    market: str = Field(..., description="市场类型")
    pinyin: Optional[str] = Field(None, description="拼音首字母")


class StockSearchResponse(BaseModel):
    """Stock search response"""

    total: int = Field(..., description="总数量")
    stocks: list[StockSearch] = Field(..., description="股票列表")


class StockHistoryPrice(BaseModel):
    """Stock historical price"""

    trade_date: datetime = Field(..., description="交易日期")
    open_price: Decimal = Field(..., description="开盘价")
    high_price: Decimal = Field(..., description="最高价")
    low_price: Decimal = Field(..., description="最低价")
    close_price: Decimal = Field(..., description="收盘价")
    volume: Decimal = Field(..., description="成交量")
    turnover: Decimal = Field(..., description="成交额")
    change_rate: Decimal = Field(..., description="涨跌幅 (%)")
